lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorRuntime.cs

```diff
@@ -164,30 +164,41 @@ public void OnNext(RuntimeStart runtimeStart)
                 catch (Exception e)
                 {
                     Utilities.Diagnostics.Exceptions.Caught(e, Level.Error, Logger);
-                    Handle(e);
+                    OnException(e);
                 }
             }
         }
 
         public void OnNext(RuntimeStop runtimeStop)
         {
             Logger.Log(Level.Info, "Runtime stop");
-            _contextManager.Dispose();
 
             if (_state == State.RUNNING)
             {
-                _state = State.DONE;
-                _heartBeatManager.OnNext();
-            }
-            try
-            {
-                _evaluatorControlChannel.Dispose();
+                const string msg = "RuntimeStopHandler invoked in state RUNNING.";
+                if (runtimeStop.Exception != null)
+                {
+                    OnException(new SystemException(msg, runtimeStop.Exception));
+                }
+                else
+                {
+                    OnException(new SystemException(msg));
+                }
             }
-            catch (Exception e)
+            else
```

Comment:
Do we need to  check if the sate is already DONE or any other state that might reach here?